pennylane/transforms/decompositions/clifford_plus_t/solovay_kitaev.py

```diff
@@ -0,0 +1,377 @@
+        )
+
+        for label in self.basis_gate:
+            op = _CLIFFORD_T_BASIS[label]
+            if qml.equal(op, last_op):
+                continue
+
+            ops = node.operations.copy()
+            ops.append(op)
+
+            if self.check_nodes(ops, seqs):
+                seqs.append(ops)
+                node.children.append(TreeSet(ops, self.basis_gate, []))
+
+        return node.children
+
+    @staticmethod
+    def check_nodes(seqs1, seqs2, tol=1e-10):
+        """Check if there's a GateSet that implements the same matrix in the given sequence up to ``tol``."""
+
+        if any(seqs1.labels == seq.labels for seq in seqs2):
+            return False
+
+        # using SO(3) instead of SU(2) since KDTrees doesn't support complex datatype
+        node_points = qml.math.array([qml.math.flatten(seq.so3_matrix) for seq in seqs2])
+        seq1_points = qml.math.array([qml.math.flatten(seqs1.so3_matrix)])
+
+        tree = sp.spatial.KDTree(node_points)
+        dist = tree.query(seq1_points, workers=-1)[0][0]
+
+        return dist > tol
+
+
+def _unitary_bloch(mat, eps=1e-10):
+    """Computes angle and axis for any given unitary matrix."""
+
+    angle = qml.math.real(qml.math.arccos(qml.math.trace(mat) / 2))
+    sine = qml.math.sin(angle)
+
+    if sine < eps:
+        axis = [0, 0, 1]
+    else:
+        nx = -qml.math.imag((mat[1, 0] + mat[0, 1]) / (2 * sine))
+        ny = qml.math.real((mat[1, 0] - mat[0, 1]) / (2 * sine))
+        nz = qml.math.imag((mat[1, 1] - mat[0, 0]) / (2 * sine))
+        axis = [nx, ny, nz]
+
+    return axis, 2 * angle
+
+
+def _group_commutator_decompose(mat):
+    """Performs group commutator decomposition U = V' @ W' @ V'.dag @ W'.dag."""
+    # Get axis and theta for the operator.
+    axis, theta = _unitary_bloch(mat)
+
+    # The angle phi comes from eq 10 in 'The Solovay-Kitaev Algorithm' paper
+    phi = 2.0 * qml.math.arcsin(qml.math.sqrt(qml.math.sqrt((0.5 - 0.5 * qml.math.cos(theta / 2)))))
+
+    # Begin decomposition by computing the rotation matrices
+    v = qml.RX(phi, [0])
+    w = qml.RY(2 * math.pi - phi, [0]) if axis[2] > 0 else qml.RY(phi, [0])
+
+    # Get similarity transormation matrix S and S.dag
+    ud = qml.math.linalg.eig(mat)[1]
+    vwd = qml.math.linalg.eig(qml.matrix(v @ w @ v.adjoint() @ w.adjoint()))[1]
+    s = ud @ qml.math.conj(qml.math.transpose(vwd))
+    sdg = vwd @ qml.math.conj(qml.math.transpose(ud))
+
+    # Get the required matrices V' and W'
+    v_hat = s @ v.matrix() @ sdg
+    w_hat = s @ w.matrix() @ sdg
+
+    return w_hat, v_hat
+
+
+def _approximate_umat(seqs, basic_approximations):
+    """Approximates given GateSet using the TreeSet structure"""
+
+    def key(x):
+        return qml.math.linalg.norm(qml.math.subtract(x.so3_matrix, seqs.so3_matrix))
+
+    return min(basic_approximations, key=key)
+
+
+def sk_approximate_set(basis_set=None, depth=10):
```

Comment:
I'm not entirely sure what this function does. Does it create some of the possible combinations of gates from the `basis_set`? And these are then used to decompose the user unitaries?

What would be an example where someone uses this function directly?